statefulset: Fix (*fakeObjectManager).UpdatePod and error tests

gjkim42 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind cleanup

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

After reviewing stateful_set_control_test.go, I found that tests for confirming the internal error are wrong.
(ref: https://github.com/kubernetes/kubernetes/pull/112812 )

And `(*fakeObjectManager).UpdatePod` does not work correctly (this should have been filtered, if the tests had worked correctly.)

This PR
- makes `fakeObjectManager` returns an error after `(*fakeObjectManager).SetUpdateStatefulPodError` is called.
- fixes error tests

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

pravarag commented:
/assign

pravarag commented:
Hi @gjkim42, I was looking at the changes in both #112813 & #112812 and was thinking if it can be considered into one PR itself or not.

gjkim42 commented:
@pravarag 

I made #112812 to prove that the current test utility cannot catch the test cases that should be failed.

pravarag commented on pkg/controller/statefulset/stateful_set_control_test.go:
```diff
@@ -2839,6 +2844,10 @@ func newRevisionOrDie(set *apps.StatefulSet, revision int64) *apps.ControllerRev
+
 	agg, ok := err.(utilerrors.Aggregate)
-	return !ok && !apierrors.IsInternalError(err) || ok && len(agg.Errors()) > 0 && !apierrors.IsInternalError(agg.Errors()[0])
+	return !ok && apierrors.IsInternalError(err) || ok && len(agg.Errors()) > 0 && apierrors.IsInternalError(agg.Errors()[0])
```
I'm thinking if this change is really required since we'll still need to check if any of api errors are not being returned for the method `isOrHasInternalError`.

gjkim42 replied:
@pravarag 

> I'm thinking if this change is really required since we'll still need to check if any of api errors are not being returned for the method `isOrHasInternalError`.

This is quite a confusing fix...

According to the contexts and test log messages of its callers, I think the function was intended to be used to catch cases that don't throw internal errors.

However,
1. the function name does not match its functionality.
2. All callers misuse this function
https://github.com/kubernetes/kubernetes/blob/6bdda2da160043c5fcdfd25bbea9e1c1b804c9e5/pkg/controller/statefulset/stateful_set_control_test.go#L415-L417

^ The above test should fail if `scaleUpStatefulSetControl` does not return an internal error, but it succeeds if it returns nil error.

This PR fixes its implementation to match its name and fixes its use cases.

soltysh commented on pkg/controller/statefulset/stateful_set_control_test.go:
```diff
@@ -2167,6 +2167,11 @@ func (om *fakeObjectManager) GetPod(namespace, podName string) (*v1.Pod, error)
+	if om.updatePodTracker.errorReady() {
+		defer om.updatePodTracker.reset()
+		return om.updatePodTracker.err
+	}
```
Good catch - thx!

soltysh reviewed (changes requested):
> After reviewing stateful_set_control_test.go, I found that tests for confirming the internal error are wrong.
(ref: https://github.com/kubernetes/kubernetes/pull/112812 )

I looked into it and it does not seem to fail for me, so your proof isn't fully working.

gjkim42 committed:
statefulset: Fix (*fakeObjectManager).UpdatePod and error tests

- Make fakeObjectManager returns an error after
  (*fakeObjectManager).SetUpdateStatefulPodError is called.
- Fix error tests

gjkim42 commented:
/retest-required

gjkim42 commented:
https://github.com/kubernetes/kubernetes/pull/112813#issuecomment-1327572998

@soltysh 

Could you take a look?
This fixes the potential false positive test cases. See https://github.com/kubernetes/kubernetes/pull/114135/commits.

gjkim42 commented:
/priority important-longterm

gjkim42 commented:
/remove-lifecycle stale
